### PR TITLE
PubSub between gregors

### PR DESCRIPTION
**Pull request by patrickxb**
* `srvup` package to maintain alive state in mysql
* `rpc/server` maintains group of alive gregord sibling connections, publishes messages to them

r? @maxtaco

### CODE REVIEW

File: bin/vendor/github.com/keybase/gregor/rpc/server/main.go [hunk 1]
```diff
@@ -285,7 +307,58 @@ func (s *Server) consume(c context.Context, m gregor1.Message) error {
 	if err := <-retCh; err != nil {
 		return err
 	}
-	s.broadcastCh <- args
+	s.broadcastCh <- messageArgs{c: c, m: m}
+
+	select {
+	case s.publishCh <- messageArgs{c: c, m: m}:
+	default:
+		s.log.Warning("publishCh full: %d", len(s.publishCh))
```
Comment (maxtaco) [inline bin/vendor/github.com/keybase/gregor/rpc/server/main.go]: This is super bad if this happens....

File: bin/vendor/github.com/keybase/gregor/rpc/server/main.go [hunk 2]
```diff
@@ -285,7 +307,58 @@ func (s *Server) consume(c context.Context, m gregor1.Message) error {
...
+			return
+		}
+	}
+}
+
+func (s *Server) publish(marg messageArgs) error {
+	s.log.Debug("publish: %+v", marg)
+	s.groupOnce.Do(func() {
+		s.log.Debug("creating new aliveGroup")
+		s.group = newAliveGroup(s.statusGroup, s.addr.String(), s.authToken, s.clock, s.closeCh, s.log)
```
Comment (maxtaco) [inline bin/vendor/github.com/keybase/gregor/rpc/server/main.go]: Getting an externally routable host/port that we can bind to might be hard in a container.  We likely need to pass it through on the command line to gregor, and figure out a way to plumb it via ECS.

File: bin/vendor/github.com/keybase/gregor/rpc/server/main.go [hunk 3]
```diff
@@ -329,6 +402,7 @@ func (s *Server) handleNewConnection(c net.Conn) error {
 
 // ListenLoop listens for new connections on net.Listener.
 func (s *Server) ListenLoop(l net.Listener) error {
+	s.addr = l.Addr()
```
Comment (maxtaco) [inline bin/vendor/github.com/keybase/gregor/rpc/server/main.go]: see above, this probably isn't going to work...

File: bin/vendor/github.com/keybase/gregor/srvup/storage_mysql.go [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+// setClock stores a clock for use in testing.  See s.now().
+func (s *StorageMysql) setClock(c clockwork.Clock) {
+	s.clock = c
+}
+
+// UpdateServerStatus implements Storage.UpdateServerStatus.
+func (s *StorageMysql) UpdateServerStatus(group, hostname string) error {
+	var err error
+	if s.update == nil {
+		if s.clock != nil {
```
Comment (maxtaco) [inline bin/vendor/github.com/keybase/gregor/srvup/storage_mysql.go]: i had a pretty good story for this exact problem in `storage/sql_sm`.  Is there anyway to use that code here?

File: srvup/storage_mysql.go [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+func (s *StorageMysql) setClock(c clockwork.Clock) {
+	s.clock = c
+}
+
+// UpdateServerStatus implements Storage.UpdateServerStatus.
+func (s *StorageMysql) UpdateServerStatus(group, hostname string) error {
+	var err error
+	if s.update == nil {
+		if s.clock != nil {
+			s.update, err = s.db.Prepare("INSERT INTO server_status (groupname, hostname, hbtime, ctime) VALUES (?, ?, ?, ?) ON DUPLICATE KEY UPDATE hbtime=?")
```
Comment (maxtaco) [inline srvup/storage_mysql.go]: i had a pretty good story for this exact problem in storage/sql_sm. Is there anyway to use that code here?

Comment (maxtaco) [inline srvup/storage_mysql.go]: Factor our my queryBuilder class?

Comment (patrickxb) [inline srvup/storage_mysql.go]: Yes, I know.  However:

1. I was trying to keep this as independent as possible.  It currently could be used by any service we come up with and be its own separate package.
2. There are only a few queries in this package and I found it easier to read and understand to see the whole query as opposed to the query builder method.

File: bin/vendor/github.com/keybase/gregor/rpc/server/main.go (hunk 2 above)
Comment (patrickxb) [inline bin/vendor/github.com/keybase/gregor/rpc/server/main.go]: It already comes as a parameter, from gregord command line options.

Comment (maxtaco) [inline bin/vendor/github.com/keybase/gregor/rpc/server/main.go]: But that bind address might be internal to Docker, and not externally router, since Docker does some NAT'ing or whatnot :(  So we might need to pass 2 addresses, one internal, and one external

Comment (maxtaco) [inline bin/vendor/github.com/keybase/gregor/rpc/server/main.go]: (Sorry should have commented on the main file, not the vendored file...)

File: bin/vendor/github.com/keybase/gregor/rpc/server/main.go (hunk 1 above)
Comment (patrickxb) [inline bin/vendor/github.com/keybase/gregor/rpc/server/main.go]: btw, this is a diff on vendoring...you might want to look at the diff of the source.  The vendoring in bin could be out of date.

File: bin/vendor/github.com/keybase/gregor/rpc/server/main.go (hunk 2 above)
Comment (patrickxb) [inline bin/vendor/github.com/keybase/gregor/rpc/server/main.go]: What's the point of an internal to docker bind address for gregor?  Isn't this bind address the address the ELB is using to connect to gregord?

Comment (maxtaco) [inline bin/vendor/github.com/keybase/gregor/rpc/server/main.go]: I don't know, that's just how it works. seems like an extra unneeded level of complication

File: srvup/storage_mysql.go (hunk 5 above)
Comment (maxtaco) [inline srvup/storage_mysql.go]: OK

**Comment by maxtaco** [comment]
Can you `git rm` the `gregor.test` binary that's checked in/

File: bin/gregord/main.go [hunk 6]
```diff
@@ -52,6 +53,35 @@ func main() {
 		db.Close()
 	}()
 
+	mstore, err := srvup.NewStorageMysql(opts.MysqlDSN, log)
```
Comment (maxtaco) [inline bin/gregord/main.go]: Can you move this logic to a subroutine to prevent too much clutter in the  main function?

File: bin/gregord/main.go [hunk 7]
```diff
@@ -52,6 +53,35 @@ func main() {
...
+	defer statusGroup.Shutdown()
+
+	alive, err := statusGroup.Alive()
+	if err != nil {
+		// bad enough to quit:
+		log.Error("%#v", err)
+		os.Exit(3)
+	}
+	// start sending heartbeats
+	statusGroup.HeartbeatLoop(opts.BindAddress)
```
Comment (maxtaco) [inline bin/gregord/main.go]: Introduce new command-line flag for externally routable incoming IP address.

File: bin/vendor/github.com/keybase/gregor/protocol/gregor1/extras.go [hunk 8]
```diff
@@ -60,8 +60,7 @@ func (d Dismissal) MsgIDsToDismiss() []gregor.MsgID {
 	return ret
 }
 
-func (m Metadata) CTime() time.Time     { return FromTime(m.Ctime_) }
-func (m Metadata) SetCTime(t time.Time) { m.Ctime_ = ToTime(t) }
```
Comment (maxtaco) [inline bin/vendor/github.com/keybase/gregor/protocol/gregor1/extras.go]: never can quite seem to get rid of this!

**Comment by patrickxb** [comment]
I did already

File: bin/vendor/github.com/keybase/gregor/protocol/gregor1/extras.go (hunk 8 above)
Comment (patrickxb) [inline bin/vendor/github.com/keybase/gregor/protocol/gregor1/extras.go]: I didn't make this change...you're in vendor again...

File: rpc/server/alive_group.go [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+	for _, host := range alive {
+		if _, ok := a.group[host]; !ok {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+func (a *aliveGroup) update() error {
```
Comment (maxtaco) [inline rpc/server/alive_group.go]: will this function ever be called more than once in parallel?

File: rpc/server/errors.go [hunk 10]
```diff
@@ -1,10 +1,38 @@
...
 func IsSocketClosedError(e error) bool {
 	return strings.HasSuffix(e.Error(), "use of closed network connection")
 }
+
+type hostErr struct {
+	host string
+	err  error
+}
+
+type pubErr struct {
```
Comment (maxtaco) [inline rpc/server/errors.go]: can you comment what this is for?

File: rpc/server/alive_group.go (hunk 9 above)
Comment (patrickxb) [inline rpc/server/alive_group.go]: no.  one goroutine running check/update.

File: rpc/transport/transport.go [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+	}
+	t.stagedTransport = rpc.NewTransport(t.conn, rpc.NewSimpleLogFactory(t.log, t.opts), keybase1.WrapError)
+	return t.stagedTransport, nil
+}
+
+func (t *connTransport) IsConnected() bool {
+	return t.transport != nil && t.transport.IsConnected()
+}
+
+func (t *connTransport) Finalize() {
```
Comment (maxtaco) [inline rpc/transport/transport.go]: why is this needed?

File: rpc/server/publish_test.go [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+	select {
+	case <-e2.BcastSent:
+	case <-time.After(2 * time.Second):
+		t.Fatal("timeout waiting for BcastSent on s2")
+	}
+
+	// check that the client connected to s2 received a broadcast
+	if len(cli.broadcasts) != 1 {
+		t.Errorf("client broadcasts received: %d, expected 1", len(cli.broadcasts))
+	}
```
Comment (maxtaco) [inline rpc/server/publish_test.go]: can we check that the pubsub'ed message is sent all the way through to the client listening on the other gregor?

File: rpc/server/alive_group.go [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+	go a.check()
+	return a
+}
+
+func (a *aliveGroup) Publish(ctx context.Context, msg gregor1.Message) error {
+	a.RLock()
+	defer a.RUnlock()
+	perr := &pubErr{}
+	for host, conn := range a.group {
+		if err := conn.CallConsumePubMessage(ctx, msg); err != nil {
```
Comment (maxtaco) [inline rpc/server/alive_group.go]: does it make sense to do these in serial?

File: rpc/server/publish_test.go [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+}
```
Comment (maxtaco) [inline rpc/server/publish_test.go]: are there tests for servers going up and down too?  Seems like we want to beef up the tests as much as possible.

File: rpc/server/publish_test.go (hunk 12 above)
Comment (patrickxb) [inline rpc/server/publish_test.go]: That's what this is.  Message was sent to s1.  cli is connected to s2.  This tests that cli received the message.

File: rpc/transport/transport.go (hunk 11 above)
Comment (patrickxb) [inline rpc/transport/transport.go]: I just moved this file from bin/gregord.

File: rpc/server/publish_test.go (hunk 12 above)
Comment (maxtaco) [inline rpc/server/publish_test.go]: Ok, is it possible to check the msg payload too, not just the existence of the message?

File: rpc/transport/transport.go (hunk 11 above)
Comment (maxtaco) [inline rpc/transport/transport.go]: Ok, you and @mmaxim shoudl coordinate, I think this is slightly old...

File: bin/vendor/github.com/keybase/gregor/rpc/server/alive_group.go [hunk 15]
```diff
@@ -0,0 +1,197 @@
...
+			} else {
+				newgroup[host] = newcli
+			}
+		}
+	}
+
+	for host, _ := range a.group {
+		if _, exists := newgroup[host]; !exists {
+			a.log.Debug("gregord on host %q no longer alive, shutting connection down", host)
+			// cli.Shutdown()
```
Comment (maxtaco) [inline bin/vendor/github.com/keybase/gregor/rpc/server/alive_group.go]: what's going on here?

Comment (patrickxb) [inline bin/vendor/github.com/keybase/gregor/rpc/server/alive_group.go]: urgh, you are in the vendor diff again...let me fix vendoring and repush.  this is different in the repo.

Comment (patrickxb) [inline bin/vendor/github.com/keybase/gregor/rpc/server/alive_group.go]: ok, vendoring should be fixed now

File: rpc/server/main.go [hunk 16]
```diff
@@ -106,12 +121,16 @@ func NewServer(log rpc.LogOutput, broadcastTimeout time.Duration) *Server {
 		syncCh:           make(chan syncArgs),
 		consumeCh:        make(chan messageArgs),
 		broadcastCh:      make(chan messageArgs),
+		publishCh:        make(chan messageArgs, 1000),
```
Comment (maxtaco) [inline rpc/server/main.go]: can this be a parameter we can tune?

File: rpc/server/main.go [hunk 17]
```diff
@@ -285,7 +308,60 @@ func (s *Server) consume(c context.Context, m gregor1.Message) error {
...
+
+// consumePub handles published messages from other gregord
+// servers.  It doesn't store to db or publish.
+func (s *Server) consumePub(c context.Context, m gregor1.Message) error {
+	s.broadcastCh <- messageArgs{c: c, m: m}
+	return nil
+}
+
+func (s *Server) publishSpawn() {
+	for i := 0; i < 10; i++ {
```
Comment (maxtaco) [inline rpc/server/main.go]: this to, can it be tunable?

File: rpc/transport/transport.go (hunk 11 above)
Comment (mmaxim) [inline rpc/transport/transport.go]: Yeah in master this is in rpc/server, although this is a fine spot too.

**Comment by maxtaco** [comment]
This is a really good start.  It has interactions with @mmaxim's branch for parallelizing DB writes. This type of code though is pretty hard to get right, so I hesitate to think I've fully understood all of the various failure cases.

File: rpc/server/main.go [hunk 18]
```diff
@@ -285,7 +308,60 @@ func (s *Server) consume(c context.Context, m gregor1.Message) error {
 	if err := <-retCh; err != nil {
 		return err
 	}
-	s.broadcastCh <- args
+	s.broadcastCh <- messageArgs{c: c, m: m}
+
+	select {
+	case s.publishCh <- messageArgs{c: c, m: m}:
+	default:
+		s.log.Warning("publishCh full: %d", len(s.publishCh))
```
Comment (mmaxim) [inline rpc/server/main.go]: Seems like this should be an error?

**Comment by patrickxb** [comment]
Thanks.  I'm sure there will be issues that we won't see until it's running...

**Comment by maxtaco** [comment]
I just have some small comments, and the approach looks solid overall.

**Comment by patrickxb** [comment]
Thanks.

It occurred to me after @mmaxim fix w/ timeouts that these intra-gregor calls also need timeouts...I will add that as well.

File: bin/gregord/main.go (hunk 6 above)
Comment (patrickxb) [inline bin/gregord/main.go]: done.

File: rpc/server/errors.go (hunk 10 above)
Comment (patrickxb) [inline rpc/server/errors.go]: done.

File: rpc/transport/transport.go (hunk 11 above)
Comment (patrickxb) [inline rpc/transport/transport.go]: Ok, I don't have a strong preference, just wanted to reuse the code.  Sounds like @mmaxim and I had the same thought in parallel branches.

Can we cleanup after merge?

File: rpc/server/publish_test.go (hunk 12 above)
Comment (patrickxb) [inline rpc/server/publish_test.go]: done

File: rpc/server/main.go (hunk 18 above)
Comment (patrickxb) [inline rpc/server/main.go]: Good idea.  I'm returning an error now.

What do you guys think about changing `default` to:

    case <-time.After(500 * time.Millisecond):

in the event of a clogged system, that might allow more messages to get through.  But perhaps having all the consume calls waiting in that scenario is a bad idea.

File: rpc/server/main.go (hunk 16 above)
Comment (patrickxb) [inline rpc/server/main.go]: done

File: rpc/server/main.go (hunk 17 above)
Comment (patrickxb) [inline rpc/server/main.go]: done

File: rpc/server/main.go (hunk 18 above)
Comment (patrickxb) [inline rpc/server/main.go]: Nevermind.  We can just increase the buffer size.

File: bin/gregord/main.go (hunk 7 above)
Comment (patrickxb) [inline bin/gregord/main.go]: done.

File: rpc/server/alive_group.go (hunk 13 above)
Comment (patrickxb) [inline rpc/server/alive_group.go]: No, I changed to do them concurrently and with a timeout.

**Comment by patrickxb** [comment]
Ok, I had a hell of a rebase, but I think I've addressed all the PR feedback.

@maxtaco @mmaxim PTAL!

**Comment by patrickxb** [comment]
I'm adding some more tests for servers going up and down...

File: bin/vendor/github.com/keybase/gregor/rpc/server/main.go [hunk 19]
```diff
@@ -293,6 +332,54 @@ func (s *Server) runConsumeMessageMainSequence(c context.Context, m gregor1.Mess
...
 
+	select {
+	case s.publishCh <- messageArgs{c: c, m: m}:
+	default:
+		s.log.Warning("publishCh full: %d", len(s.publishCh))
+		return ErrPublishChannelFull
+	}
+	return nil
+}
+
```
Comment (mmaxim) [inline bin/vendor/github.com/keybase/gregor/rpc/server/main.go]: It might make sense to pull this guy into a function called `publishConsumeMessage`. Max wanted this function to really make the whole routine very explicit, which is why I make the broadcast and storage version.

Comment (mmaxim) [inline bin/vendor/github.com/keybase/gregor/rpc/server/main.go]: It if wasn't clear, I'm referring to the code above this comment

Comment (maxtaco) [inline bin/vendor/github.com/keybase/gregor/rpc/server/main.go]: 👍

Comment (patrickxb) [inline bin/vendor/github.com/keybase/gregor/rpc/server/main.go]: got it.  yeah, I saw that as I was rebasing.  Will change now...

File: bin/vendor/github.com/keybase/gregor/rpc/server/main.go [hunk 20]
```diff
@@ -293,6 +332,54 @@ func (s *Server) runConsumeMessageMainSequence(c context.Context, m gregor1.Mess
...
+// consumePub handles published messages from other gregord
+// servers.  It doesn't store to db or publish.
+func (s *Server) consumePub(c context.Context, m gregor1.Message) error {
+	s.broadcastCh <- messageArgs{c: c, m: m}
+	return nil
+}
+
```
Comment (mmaxim) [inline bin/vendor/github.com/keybase/gregor/rpc/server/main.go]: I wonder if we should just pick a name for "publish", either but mixing up "publish" and "pub" might not be the best play. We could think about renaming this to `consumePublish`, since "pub" can kind of map to a bunch of different concepts on first read.

Comment (mmaxim) [inline bin/vendor/github.com/keybase/gregor/rpc/server/main.go]: Also I realize this is the vendored code now, it just comes first in the display!

File: bin/vendor/github.com/keybase/gregor/rpc/server/sibconn.go [hunk 21]
```diff
@@ -0,0 +1,133 @@
...
+		return nil, err
+	}
+
+	return s, nil
+}
+
+func (s *sibConn) CallConsumePubMessage(ctx context.Context, msg gregor1.Message) error {
+	ctx, cancel := context.WithTimeout(ctx, s.timeout)
+	defer cancel()
+	ic := gregor1.IncomingClient{Cli: s.conn.GetClient()}
```
Comment (mmaxim) [inline bin/vendor/github.com/keybase/gregor/rpc/server/sibconn.go]: As I discovered after talking to strib, you could just keep around a client on `sibConn`, you just have to grab it in `connect` rather than `OnConnect`

**Comment by maxtaco** [comment]
👍 ready when you are

**Comment by mmaxim** [comment]
Same!

**Comment by jzila** [comment]
This seems to be breaking KBFS docker tests. Here's an example: https://travis-ci.com/keybase/kbfs/jobs/39968094#L798
